blob/service.go

```diff
@@ -347,127 +354,204 @@ func (s *Service) retrieve(
+				return nil, nil, err
+			}
+			blobLen := shares.SparseSharesNeeded(sequenceLength)
 
-		proofs = append(proofs, row.Proof)
-		index := row.Proof.Start()
-
-		for {
-			var (
-				isComplete bool
-				shrs       []shares.Share
-				wasEmpty   = sharesParser.isEmpty()
-			)
-
-			if wasEmpty {
-				// create a parser if it is empty
-				shrs, err = sharesParser.set(rowIndex*len(header.DAH.RowRoots)+index, appShares)
-				if err != nil {
-					if errors.Is(err, errEmptyShares) {
-						// reset parser as `skipPadding` can update next blob's index
-						sharesParser.reset()
-						appShares = nil
-						break
-					}
-					return nil, nil, err
-				}
-
-				// update index and shares if padding shares were detected.
-				if len(appShares) != len(shrs) {
-					index += len(appShares) - len(shrs)
-					appShares = shrs
-				}
+			endShareIndex := currentShareIndex + blobLen
+			if endShareIndex > len(rowsShares) {
+				// this blob spans to the next row which has a namespace > requested namespace.
+				// this means that we can stop.
+				return nil, nil, ErrBlobNotFound
+			}
+			// convert the blob shares to app shares
+			blobShares := rowsShares[currentShareIndex:endShareIndex]
+			appBlobShares, err := toAppShares(blobShares...)
+			if err != nil {
+				return nil, nil, err
 			}
 
-			shrs, isComplete = sharesParser.addShares(appShares)
-			// move to the next row if the blob is incomplete
+			// parse the blob
+			sharesParser.length = blobLen
+			_, isComplete := sharesParser.addShares(appBlobShares)
 			if !isComplete {
-				appShares = nil
-				break
+				return nil, nil, fmt.Errorf("expected the shares to construct a full blob")
 			}
-			// otherwise construct blob
 			blob, err := sharesParser.parse()
 			if err != nil {
 				return nil, nil, err
 			}
 
-			if sharesParser.verify(blob) {
-				return blob, &proofs, nil
-			}
+			// number of shares per EDS row
+			numberOfSharesPerEDSRow := squareSize * 2
+			// number of shares from square start to namespace start
+			sharesFromSquareStartToNsStart := inclusiveNamespaceStartRowIndex * numberOfSharesPerEDSRow
+			// number of rows from namespace start row to current row
+			rowsFromNsStartToCurrentRow := currentShareIndex / squareSize
+			// number of shares from namespace row start to current row
+			sharesFromNsStartToCurrentRow := rowsFromNsStartToCurrentRow * squareSize * 2
+			// number of shares from the beginning of current row to current share
+			sharesFromCurrentRowStart := currentShareIndex % squareSize
+			// setting the index manually since we didn't use the parser.set() method
+			blob.index = sharesFromSquareStartToNsStart +
+				sharesFromNsStartToCurrentRow +
+				sharesFromCurrentRowStart
+
+			if blob.Namespace().Equals(namespace) && sharesParser.verify(blob) {
+				// now that we found the requested blob, we will create
+				// its inclusion proof.
+				inclusiveBlobStartRowIndex := inclusiveNamespaceStartRowIndex + currentShareIndex/squareSize
```

Comment:
Do we need this calculation after we've calculated the `blob.Index`? can't it be calculated this way: `blob.Index/(squareSize*2)`?

Reply:
yes, want me to update?

```suggestion
				inclusiveBlobStartRowIndex := blob.index / (squareSize * 2)
```

Reply:
yes, please